Newsletter Page

JessSumner opened the pull request:
Why:
The rescue distributes a monthly flyer and a newsletter. It would be nice to have a library of these documents and a place where visitors can subscribe to a mailing list.

This PR:
Contains a controller that manages accessing the documents from asset folders and storing them as an array and set to a variable.
The view then iterates through these arrays and displays a link to the document as a thumbnail preview of the document.
This PR states that there will be links to a mailing list and a contact us. These links are not currently hooked up to anything.

JessSumner committed:
Generated controller and views for newsletters

Newsletters controller has two methods: index and show.
Created views for both index and show.
Created routes for both index and show.
Linked up the navigation bar link to the index page for newsletters.

JessSumner committed:
Displaying monthly flyers

Ammended fliers to flyers. Decided that flyers is more common and have decided this is how I am going to refer to them. All files have been ammended to use this spelling.
Add 13 flyers to an assets file.
Created a FLYERS constant that converts all the flyers file names to strings.
The FLYERS constant can be called and iterated through to render the view.
N.B the flyers are being accessed but not loading correctly in the view.

JessSumner committed:
Index loop is working

In the index view is calling the FLYERS constant and looping through them. For each one it creates a link and displays it's name.
Currently the link is not working.
There is no styling done on the list.

JessSumner committed:
Fixed the links to the flyer pdfs

Why:
The links were appearing but producing an error when clicked.

This commit:
Removes newsletters from the link path. This makes it work.
Removed the class from the index method as it is within the Newsletters class so it can see the constant.

JessSumner committed:
Removed the dependency on the FLYERS constant

Why:
In order to add a new flyer to the website it would need to be added to the assets file but also the name would need to be put into the constant. It would need to be put in the correct order as well. It would be nice to only have to add the file to the assets folder.

This commit:
Includes the code to perform the following actions.
1 - The contents of the monthly_flyers folder within the assets folder are retrieved.
2 - Select only files with .pdf in the name
3 - Sort the files by date.
4 - Remove the .pdf file extension

JessSumner committed:
Added thumbnails for the flyers

Why:
The page will look better with thumbnails of the flyers that can be clicked to enlarge.

This commit:
Adds png files of all the flyers to the assets folder. The png has the same name as the corresponding pdf. Therefore the same methods can be used to access the png as the pdf.
Formatted the text and thumbnails on the page.

JessSumner committed:
Quarterly newsletters added

Why:
Any rat rescue also creates a newsletter periodically. It would be nice to display these too.

This commit:
Adds the newsletters I have access to to the assets folder.
Repeats the logic used to display the flyers for the newsletters.

JessSumner committed:
Refactor of the newsletters controller

Why:
Both news and flyers were calling their own set of the same methods.

This commit:
The method chain is the same and the other than the first and final methods they do not need to know which folder they are dealing with. The first method searches through the directory of a folder which is provided by an instance variable defined in the final method. The final method is setting the array of either flyers or news and so can specify which folder the method chain should begin.

JessSumner committed:
Added subscribe and email us link

Why:
Want it to be easy for visitors to subscribe to a mailing list. Also want visitors who want to post the flyer to contact us to let us know.

This commit:
Adds a subscribe button.
Adds an email us link.
Neither of these links are hooked up yet.
Centered the text as this page doesn't have blocks of text and justify wasn't looking quite right.

JessSumner commented:
@MattMSumner @halogenandtoast @kenyonj I would be very interested in your feedback on this PR. 

I am very excited by the refactor for the controller. However I am curious as to whether it can be improved upon.

I am also interested to know whether you feel a pdf and png is necessary for the flyers. A png can be used for the thumbnail but also open to be viewed more clearly. The pdf is redundant other than it allows the user to download it if they want to display it for the rescue. I also feel it looks better. However it is two files for every item. The quarterly newsletters need both as the pdf is multiple pages and the png is just the first page.

Thank you very much.

JessSumner committed:
Hound corrections

Deleted newsletter_controller_test.rb
Added space before opening block and after commas.
Removed space after last block parameter.
Added more descriptive class names so that I can avoid using class selectors with an element.

halogenandtoast commented on app/controllers/newsletters_controller.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def file_contents
+    Dir.entries("app/assets/newsletters/#{@file}")
```
I prefer not to use instance variables this way in the controller, instead opting to pass them down. However this seems like a lot of responsibility for the controller in the first place. I'd suggest pulling out another object to handle this. Maybe something like `NewsletterFinder.new("news").find` and `NewsletterFinder.new("flyers").find`

halogenandtoast commented on app/views/newsletters/index.html.erb:
```diff
@@ -0,0 +1,45 @@
+    Want to be kept up to date with the latest news from Any Rat Rescue? 
+  </p>
+  <button class="subscribe"><%= link_to "Subscribe", "#" %></button>
+  <h1>
```
This h1 feels out of place in this `div` perhaps this should be a `header` tag instead and then you could change the div below to a `main` tag

halogenandtoast replied:
I mean wrap the content below in a `main` tag

JessSumner replied:
@halogenandtoast I agree. The text I am referring to are subheadings and so don't really need to be h1. On looking at my other pages I realized that I was using h2 tags. I have changed it this way. I hope that makes sense.

halogenandtoast commented:
> I am also interested to know whether you feel a pdf and png is necessary for the flyers. A png can be used for the thumbnail but also open to be viewed more clearly. The pdf is redundant other than it allows the user to download it if they want to display it for the rescue. I also feel it looks better. However it is two files for every item. The quarterly newsletters need both as the pdf is multiple pages and the png is just the first page.

Hard to say. If they already generate both then it seems fine to use both. If not I would go with whatever most closely matches what they currently do.

JessSumner committed:
Removed the instance variables defined in the controller.

It is good practice to avoid using instance variables within the controller.

I have updated the methods to take an argument and therefore eliminate the need for defining the instance variable.

JessSumner committed:
Created Newsletter Finder Model

Why:
The Newsletter controller was very logic heavy. It handled and had access to more information than it needed.

How:
Created a model NewsletterFinder.
Logic to find files, sort and format the names has been moved to the model.
The controller can call a new NewsletterFinder object and call the final method (either news or flyers) which will then proceed through the logic in the model.

JessSumner commented:
@halogenandtoast Thank you for the comments on the controller. I agreed immediately but it took me a little while to realize what needed to be done. I have added two commits which encompasses your suggestion. The first removes the instance variable and the second moves the logic. I feel happier about the code. Looking in, is the code easier to read and understand?

JessSumner committed:
Corrections

Deleted the newsletters/show.html.erb file. This file was added during the initial planning. It is not being used as each flyer opens as a pdf so does not need a view.
Removed an empty line from the migration file.
Changed the h1 tags to h2 tags. The text in these tags are sub headings not titles. This is flows better with the style of the overall document.

halogenandtoast commented on app/controllers/newsletters_controller.rb:
```diff
@@ -0,0 +1,6 @@
+class NewslettersController < ApplicationController
+  def index
+    @flyers = NewsletterFinder.new.flyers
```
I would do this a bit differently. It's hard to explain why, but I think it's preferable to do the following:

``` ruby
@flyers = NewsletterFinder.new("flyers")
```

Then in `NewsletterFinder` you could `include Enumerable` which I'll let you explore on your own, but essentially if you look at your `news` and `flyers` functions they are so very similar that if you instead instantiated your object with that string then you could have a single method instead and your class would be more versatile (if you added a third type later on you wouldn't have to change your class - This is known as the Open/Closed Principle).

JessSumner replied:
I have heard of this principle. As it happens I have just started the chapter in POODR where she explains it. I will continue reading and try to apply it. 
Hopefully I can come back to you with something cleaner.

MattMSumner commented on app/views/newsletters/index.html.erb:
```diff
@@ -0,0 +1,45 @@
+
+<div class="newsletter-text">
+  <p>
+    Own a business and want to post our flyer? We appreciate the support and would love to hear from you. Please <%= link_to "email us", "#" %> and tell us your story.
```
How would you feel about breaking up this line?

kenyonj commented on app/views/newsletters/index.html.erb:
```diff
@@ -0,0 +1,45 @@
+  <p>
+    Want to be kept up to date with the latest news from Any Rat Rescue? 
+  </p>
+  <button class="subscribe"><%= link_to "Subscribe", "#" %></button>
```
Since you have `bourbon` in this project, can you just use the `button` class on the `link_to` instead of wrapping it in a button element?

JessSumner replied:
@kenyonj So I have changed the line to be 

``` erb
<%= link_to "Subscribe", "#", class: "button" %>
```

but that just gives me a link. It no longer looks like a button. Am I missing something?

kenyonj replied:
Sorry, it looks like the button mixin from bourbon has been deprecated (see here: http://bourbon.io/docs/#buttons ). You could add [bitters](http://bitters.bourbon.io/) to the project and that will give you that button style. 

Usually I try and just use `link_to`s and just style them as buttons.

MattMSumner replied:
I think this is using bitters as well? https://github.com/JessSumner/any_rat_rescue/tree/master/app/assets/stylesheets/base

JessSumner replied:
``` erb
<%= link_to "Subscribe", "#", {style: "button"} %>
<%= link_to "Subscribe", "#", {class: "button"} %>
```

@kenyonj I'm curious as I still haven't managed to make it work. I tried both of these. Any ideas?

kenyonj commented on db/migrate/20160203152552_create_newsletter_finders.rb:
```diff
@@ -0,0 +1,7 @@
+class CreateNewsletterFinders < ActiveRecord::Migration
+  def change
+    create_table :newsletter_finders do |t|
+      t.timestamps null: false
```
Is there a reason that there is a table for this model? Could it just be a [PORO](http://blog.jayfields.com/2007/10/ruby-poro.html)?

JessSumner replied:
@kenyonj Yes it definitely can. Thank you for the catch.

JessSumner committed:
Removed Active Record inheritance for the model

Why:
This model does not need a database associated with it. It is an object to contain logic.

This commit:
Removes the inheritance.
Rolls back the migration.
Deletes the migration file.

Everything still works as it should after a server restart!

kenyonj commented on app/models/newsletter_finder.rb:
```diff
@@ -0,0 +1,27 @@
+class NewsletterFinder
```
It looks like there is the concept of `newsletters` and `flyers`. Have you thought of separating these out? It would not only solve your "open/closed principle" violation as @halogenandtoast suggested, but also make it clear what is being duplicated that you could possibly extract into a new parent class.

JessSumner replied:
@kenyonj I like this idea too. Thank you. I ended up using require Enumerable. My brain hurt after it but I like the result.

JessSumner committed:
Utilizing Enumerable and delegate

Why:
There are a lot of dependencies between NewsletterController and NewsletterFinder. NewsletterController knows there is an object NewsletterFinder and it also knows two of the methods that can be called. NewsletterFinder is also restricted to retrieving monthly_flyers and news. It would be good to open up this class to be able to accept changes to either of these folders or an addition of another.

How:
1) For NewsletterController instead of requesting a specific method on a new instance of NewsletterFinder. Request the NewsletterFinder and pass an argument. The argument represents the folder which holds the documents desired.
2) NewsletterFinder needs an initialize method which accepts an argument. This removes all responsability of the NewsletterFinder class to know what the folder is. This allows changes to the folder name to be made easily.
3) This removes the methods flyers and news.
4) The methods flyers and news had a dual purpose. They defined the folder which was to be passed through the methods and it also removed the file extension from the file names.
5) The method to remove the file extension is still required but only once.
6) Making these changes results in an error. Undefined method each for @flyers.each in the view. Each is a method which is included in the Enumerable Module. The array class includes the Enumerable Module and therefore if you call .each on an array you get the expected behaviour. The NewsletterFinder is not an array and @flyers is an instance of NewsletterFinder. It does not include the Enumerable Module and therefore any methods such as each, map, select and sort are not supported.
7) require Enumerable. This is therefore included at the top of the NewsletterFinder class
8) require Enumerable then requires you to define each. Having each defined allows other methods such as map to be used.
9) In this example each needs to yield each flyer or news document. Instead of defining it myself I delegate the responsability to an object I know to be an array newsletter_list.
10) @newsletter_list has been added as everytime .each is called on an instance of NewsletterFinder such as on @flyers it will run through the whole chain of methods. Defining this instance variable stores it so that the first time it is called all the methods will run, however the second time there is a memory.
11) Making methods in the NewsletterFinder private. The reason for making the NewsletterFinder class was to remove the responsability from other objects. The NewsletterFinder's sole purpose is to retrieve files from a given file for other objects to then use. Instead of calling one of the methods in the chain only the object NewsletterFinder should be called.

JessSumner committed:
Changed folder name

Why:
I was using "news" for the folder name and it's contents.

How:
Replaced news in the folder name, the NewsletterController and the index view.

Lack of side-effect:
Nothing needed to be changed in the NewsletterFinder object. This was the goal of making this object and separating the logic.

JessSumner commented:
@halogenandtoast How's this? You have made me learn a lot by suggesting this! Thank you. I'm sure the commit message can be cut down. I was re-working through it in my head.

MattMSumner commented:
This LGTM :+1:

halogenandtoast commented on app/models/newsletter_finder.rb:
```diff
@@ -0,0 +1,31 @@
+
+  def newsletters_with_file_extension
+    unsorted_newsletters.sort do |date_1, date_2|
+      Date.parse("1-#{date_2}") <=> Date.parse("1-#{date_1}")
```
This seems a little brittle, why is the "1-" necessary?

JessSumner replied:
![screen shot 2016-02-04 at 9 39 28 pm](https://cloud.githubusercontent.com/assets/11034063/12836530/da334068-cb87-11e5-8d59-0d318add53cd.png)
@halogenandtoast Without the 1- it does not accept the argument as a date. The files are saved as jun-15 for example and I mimicked this in the example in irb.

halogenandtoast replied:
@JessSumner won't this be a problem when they have more than 1 years worth of content?

JessSumner replied:
@halogenandtoast It isn't a problem. The one is just representing the day. I already have two flyers for January and two for December up already. I didn't think of the day when I named the files as I only associate them with the month. I could change the file names to include the 1 so this can be removed but is it worth it?

halogenandtoast replied:
Ah right, I don't frequently parse dates in that format so for some reason I thought the 15/16 in your example were days and 1 was a year, but this makes sense. I wouldn't worry about changing anything.

halogenandtoast commented:
@JessSumner The class looks great, just have one question about the sorting.
